File: api.integration.tests/Http.cs
using common;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using System;

namespace api.integration.tests;

internal static class HttpModule
{
    public static string ApiClientKey { get; } = nameof(ApiClientKey);

    public static void ConfigureHttpClient(IHostApplicationBuilder builder)
    {
        builder.Services.AddServiceDiscovery()
                        .ConfigureHttpClientDefaults(http =>
                        {
                            http.AddStandardResilienceHandler();
                            http.AddServiceDiscovery();
                        })
                        .AddHttpClient(ApiClientKey, client =>
                        {
                            var apiConnectionName = builder.Configuration.GetValueOrThrow("ASPIRE_API_CONNECTION_NAME");

                            client.BaseAddress = new($"https+http://{apiConnectionName}");
                            client.Timeout = TimeSpan.FromMinutes(3);
                        });
    }
}

File: api/code/api.integration.tests/Http.fs
[<RequireQualifiedAccess>]
module api.integration.tests.Http

open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Hosting
open OpenTelemetry.Metrics
open OpenTelemetry.Trace
open System
open System.Net.Http
open common

let configureOpenTelemetryMetrics (metrics: MeterProviderBuilder) =
    metrics.AddHttpClientInstrumentation() |> ignore

let configureOpenTelemetryTracing (tracing: TracerProviderBuilder) =
    tracing.AddHttpClientInstrumentation() |> ignore

let private configureHttpClientBuilder (builder: IHttpClientBuilder) =
    builder.AddStandardResilienceHandler() |> ignore
    builder.AddServiceDiscovery() |> ignore

let apiIdentifier = "api"

let private configureApiHttpClient configuration (client: HttpClient) =
    let connectionName =
        Configuration.getValueOrThrow configuration "API_CONNECTION_NAME"

    client.BaseAddress <- Uri($"https+http://{connectionName}")

let configureBuilder (builder: IHostApplicationBuilder) =
    builder.Services
        .AddServiceDiscovery()
        .ConfigureHttpClientDefaults(configureHttpClientBuilder)
        .AddHttpClient(apiIdentifier, configureApiHttpClient builder.Configuration)
    |> ignore
